### PR TITLE
No U Turn Sampler

### PR DESCRIPTION
**Pull request by khalibartan**
This PR Deals with implementing following two samplers:
- NUTS
- NUTS with dual averaging

### CODE REVIEW

**Comment by khalibartan** [comment]
@ankurankan The `_build_tree` method for NUTS and NUTS with dual averaging is slightly different for now I have wrote different code for both. Is that okay, or should I try to accommodate things in a single method with if else etc.

File: pgmpy/inference/continuous/sampling.py [hunk 1]
```diff
@@ -591,3 +592,231 @@ def generate_sample(self, initial_pos, num_adapt, num_samples, trajectory_length
...
+        samples[0] = tuple(initial_pos)
+        position_m = initial_pos
+
+        for i in range(1, num_samples):
+            # Genrating sample
+            position_m = self._sample(position_m, stepsize)
+            samples[i] = position_m
+
+        if HAS_PANDAS == True:
+            return pd.DataFrame.from_records(samples)
```
Comment (yashu-seth) [inline pgmpy/inference/continuous/sampling.py:836]: @khalibartan There could be case when the user would want a np.recarry output despite having pandas installed. May be we can have an optional parameter to ensure this. What do you think?

Comment (khalibartan) [inline pgmpy/inference/continuous/sampling.py:836]: @yashu-seth Earlier in HMC I have done the same, but @ankurankan told me to return pandas.DataFrame if available (i.e current way of returning ).

Comment (yashu-seth) [inline pgmpy/inference/continuous/sampling.py:836]: @khalibartan Okay. I have not gone through all the methods. Can you tell me if this output would be used in some other pgmpy method?

**Comment by khalibartan** [comment]
@ankurankan Please take a look at things. I haven't added tests for `_build_tree` method. Should I add tests for it? I'll make some changes in it after PR about joint gaussian distribution is merged.

File: pgmpy/inference/continuous/sampling.py [hunk 2]
```diff
@@ -49,12 +51,12 @@ class HamiltonianMC(object):
     >>> samples = sampler.sample(initial_pos=np.array([1, 1]), num_samples = 10000,
     ...                          trajectory_length=2, stepsize=0.4)
     >>> samples
-    array([(5e-324, 5e-324), (-2.2348941964735225, 4.43066330647519),
-           (-2.316454719617516, 7.430291195678112), ...,
-           (-1.1443831048872348, 3.573135519428842),
-           (-0.2325915892988598, 4.155961788010201),
-           (-0.7582492446601238, 3.5416519297297056)], 
-          dtype=[('x', '<f8'), ('y', '<f8')])
+    rec.array([(1.0, 1.0), (-3.1861687131079086, 3.7940994520145654),
```
Comment (ankurankan) [inline pgmpy/inference/continuous/sampling.py:54]: @khalibartan Format this properly.

File: pgmpy/inference/continuous/sampling.py [hunk 3]
```diff
@@ -591,3 +593,629 @@ def generate_sample(self, initial_pos, num_adapt, num_samples, trajectory_length
...
+    >>> from pgmpy.models import JointGaussianDistribution as JGD
+    >>> import numpy as np
+    >>> mean = np.array([1, 2, 3])
+    >>> covariance = np.array([[4, 0.1, 0.2], [0.1, 1, 0.3], [0.2, 0.3, 8]])
+    >>> model = JGD(['x', 'y', 'z'], mean, covariance)
+    >>> sampler = NUTS(model=model, grad_log_pdf=None, simulate_dynamics=LeapFrog)
+    >>> samples = sampler.sample(initial_pos=np.array([0.1, 0.9, 0.3]), num_samples=20000,stepsize=0.4)
+    >>> samples
+    rec.array([(0.1, 0.9, 0.3),
+     (-0.27303886844752756, 0.5028580705249155, 0.2895768065049909),
```
Comment (ankurankan) [inline pgmpy/inference/continuous/sampling.py:632]: @khalibartan Format this properly.

**Comment by ankurankan** [comment]
@khalibartan This PR looks good. I am merging this right now. And please create a new PR for the remaining tests and other changes.

File: pgmpy/inference/continuous/sampling.py (hunk 3 above)
Comment (khalibartan) [inline pgmpy/inference/continuous/sampling.py:632]: @ankurankan This is the actual output representation. I have copied it.

**Pull request merged**